src/lib/handler_monitor.cc

```diff
@@ -183,14 +204,75 @@ int handler_monitor::_down() {
 }
 
 int handler_monitor::_up() {
-	if (this->_down_state >= this->_monitor_threshold) {
+	if (this->_down_count >= this->_monitor_threshold) {
 		log_info("node seems up -> dispatch node up event", 0);
 		this->_cluster->request_up_node(this->_node_server_name, this->_node_server_port);
 	}
-	this->_down_state = 0;
+	this->_down_count = 0;
 
 	return 0;
 }
+
+void handler_monitor::_process_node_map_version(const stats_results& results) {
+	stats_results::const_iterator result = results.find_by_name("flare_node_map_version");
+	if (result == results.end()) {
+		return;
+	}
+	uint64_t version;
+	try {
+		version = boost::lexical_cast<uint64_t>(result->value);
+	} catch (boost::bad_lexical_cast&) {
+		log_debug("invalid version");
+		return;
+	}
+	this->_update_node_map_version_match_count(version);
+}
+
+handler_monitor::node_status handler_monitor::_get_node_status(const stats_results &results) {
+	stats_results::const_iterator result = results.find_by_name("flare_node_status");
+	if (result == results.end()) {
+		return node_status_not_found;
+	}
+	if (result->value != "OK") {
+		return node_status_result_ng;
+	}
+	return node_status_result_ok;
+}
+
+void handler_monitor::_update_node_map_version_match_count(uint64_t version) {
+	if (this->_cluster->get_node_map_version() == version) {
+		this->_node_map_version_mismatch_count = 0;
+		return;
+	}
+
+	this->_node_map_version_mismatch_count++;
+
+	static const double mismatch_threshold_sec = 30;
+	const uint64_t mismatch_threshold = ceil(mismatch_threshold_sec / this->_monitor_interval);
+	if (this->_node_map_version_mismatch_count >= mismatch_threshold) {
+		log_err(
+			"node_map_version seems mismatched (current_index_server:%"PRIu64", node(%s:%d):%"PRIu64")",
+			this->_cluster->get_node_map_version(),
+			this->_node_server_name.c_str(),
+			this->_node_server_port,
+			version
+		);
+	}
+}
+
+int handler_monitor::_request_stats(stats_results &results) {
+	boost::shared_ptr<op_stats> op(new op_stats(this->_connection));
```

Comment:
Why do you use shared_ptr even though this pointer is not shared with other objects?